/*
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Data models for describing a vulnerability detection report.
syntax = "proto3";

package tsunami.proto;

import "google/protobuf/timestamp.proto";
import "network_service.proto";
import "reconnaissance.proto";
import "vulnerability.proto";

option java_multiple_files = true;
option java_outer_classname = "DetectionProtos";
option java_package = "com.google.tsunami.proto";
option go_package = "github.com/google/tsunami-security-scanner/proto";

// Status of the vulnerability detection result.
enum DetectionStatus {
  // Unspecified status.
  DETECTION_STATUS_UNSPECIFIED = 0;
  // Target is not vulnerable.
  SAFE = 1;
  // Target appears to be vulnerable (e.g. because running version is
  // vulnerable), but couldn't be verified.
  VULNERABILITY_PRESENT = 2;
  // Target is vulnerable and the detector successfully verified the
  // vulnerability.
  VULNERABILITY_VERIFIED = 3;
}

// Full report about a detected vulnerability.
message DetectionReport {
  // Information about the scanned target.
  TargetInfo target_info = 1;

  // Information about the scanned network service.
  NetworkService network_service = 2;

  // Time when the vulnerability was detected.
  google.protobuf.Timestamp detection_timestamp = 3;

  // Status of the detection result.
  DetectionStatus detection_status = 4;

  // Full details about the detected vulnerability.
  Vulnerability vulnerability = 5;
}

message DetectionReportList {
  repeated DetectionReport detection_reports = 1;
}
